### DISCUSSION TITLE
py::object: docs / sources?

### DISCUSSION THREAD

**Discussion by RobG-LHind** [pybind/pybind11#4252] [Q&A]
Hello, I am new to pybind11. 
I tried to find answers in https://pybind11.readthedocs.io/en/stable/.... and on stackoverflow, but I may have overlooked the obious: I do not find the sources of py::object (declaration, implementation) or a detailed documentation of py::object (attributes, methods). 
Can someone give me a link?

aim:
In a C++ function/method I would like to modify a python-object of a non standard type (cv2.cuda_GpuMat), but I do not find any detailed information about py::object . 

How do I check the type of the py::object / PyObject?
How do I get (/iterate over) pointer/references to attributes of the py::pbject / PyObject?

use case:
C++ library cppXXX exists.
Incomplete python wrapper pyXXX of cppXXX exists.
    struct XXX::YYY{ void *data;}};
    Python class XXX_YYY exists, there is *data (among other things) inside.

I would like to write a C++ function 
    void coPyYYY(XXX:YYY &cyyy, py::object &pyyy){
      // copy cyyy.data to pyyy.data  
      // increase refcounter
      // choose right policy for destructing

**Comment by jiwaszki** [comment]
Here https://pybind11.readthedocs.io/en/stable/advanced/pycpp/object.html#available-wrappers and later here https://pybind11.readthedocs.io/en/stable/reference.html#_CPPv46object

If you want more details look here:
https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L309

> How do I check the type of the py::object / PyObject?

You probably look for this(?):
https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L144
.. or this:
https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L187
... or even this:
https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L740-L748
There is a lot of them to choose from and combine in different scenarios.

> How do I get (/iterate over) pointer/references to attributes of the py::pbject / PyObject?

I am aware that you are able to get them by specific keys, however not quite sure if it is possible to iterate over them:
https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L101-L111
... and docs:
https://pybind11.readthedocs.io/en/stable/reference.html?highlight=object#python-built-in-functions

Maybe some kind of CPython call can do this? Or maybe try working out with interpreter calls in pybind's layer (like getting `__dir__`)?

Hope it helps, cheers!

**Reply by RobG-LHind** [reply]
Hello Jan, 
Thank you very much for your answer. 
The links you provided have a lot of information about py::object. 
After reading lots of paragraphs, I get some steps forward, but main problems persist.

How do I get the (string-)type information about the PyObject/py::object?
I can check `PyObject_HasAttrString`, but how do I get the AttrString of a given PyObject?

How do I access the pointer to the values stored inside a PyObject (`PyLong_AsVoidPtr`)?

Cheers
Robert


    int modify(cv::cuda::GpuMat* cVal, py::object &pobj)
 
     {

        py::handle ptype = pobj.get_type(); //?? ...

        PyObject *pobject = pobj.ptr();

        auto typeCheck = PyObject_HasAttrString(pobject, "cv2.cuda_GpuMat");

        if (!typeCheck) 

          return(0);

        PyObject *pobject1 = PyObject_GetAttrString( pobject, "cv2.cuda_GpuMat");

        cv::cuda::GpuMat *pyVal = (cv::cuda::GpuMat *)PyLong_AsVoidPtr(pobject1);

        // *pyVal = *cVal;

        return(1);

      }

**Reply by jiwaszki** [reply]
Did you try casting `py::object` to `py::handle`? Or just using it directly? You will be able to retrieve a pointer to `PyObject` and later you might cast it to OpenCV's class.

Ref: https://pybind11.readthedocs.io/en/stable/reference.html?highlight=handle#_CPPv4NK6handle3ptrEv
